optax/losses/_classification.py

```diff
@@ -430,3 +430,48 @@ def ctc_loss(
       logits, logit_paddings, labels, label_paddings,
       blank_id=blank_id, log_epsilon=log_epsilon)
   return per_seq_loss
+
+
+
+def sigmoid_focal_loss(
+    logits:  chex.Array,
+    labels:  chex.Array,
+    alpha: Optional[float] = None,
+    gamma: float = 2,
+) ->  chex.Array:
+  """Compute a sigmoid focal loss.
+  The focal loss is a re-weighted cross entropy for unbalanced problems.
+  Use this loss function if classes are not mutually exclusive.
+  See `sigmoid_binary_cross_entropy` for more information.
+
+  References:
+    Lin et al. 2018. https://arxiv.org/pdf/1708.02002.pdf
+  
+  Args:
+    logits: Array of floats. The predictions for each example.
+      The predictions for each example.
+    labels: A float array. Labels and logits must have
+      the same shape. The label array must contain the binary
+      classification labels for each element in the data set
+      (0 for the out-of-class and 1 for in-class).
+    alpha: (optional) Weighting factor in range (0,1) to balance
+      positive vs negative examples. Default None (no weighting).
+    gamma: Exponent of the modulating factor (1 - p_t) to
+      balance easy vs hard examples.
+
+  Returns:
+    A loss value array with a shape identical to the logits and target
+    arrays.
+  """
+  chex.assert_type([logits], float)
+  labels = labels.astype(logits.dtype)
+  # see also the original paper's implementation at:
+  # https://github.com/facebookresearch/fvcore/blob/main/fvcore/nn/focal_loss.py
+  p = jax.nn.sigmoid(logits)
+  ce_loss = sigmoid_binary_cross_entropy(logits, labels)
+  p_t = p * labels + (1 - p) * (1 - labels)
+  loss = ce_loss * ((1 - p_t) ** gamma)
+  if alpha:
```

Comment:
Won't this prevent this from being JIT'able? I ran it briefly and it errored on this. It also seems like in the case that alpha=0.0, the loss is the same - so can we just set alpha=0.0 by default and drop the conditional?

Reply:
Dear @the-xentropy, thank you for reviewing the code. JIT support is essential. Unfortunately, `alpha=0` and `alpha=None` are not the same. Therefore we can't drop the conditional, but we can use `jax.lax.cond`, [7368964](https://github.com/google-deepmind/optax/pull/599/commits/7368964be9b2713600870eb88d24f3d741a6ef9c) does this. I hope it addresses your concern.

Reply:
thanks for catching this @the-xentropy and @v0lta for the fix. Could you also add a test that JITs the function so it doesn't break in the future?

Reply:
Sure, I think `@chex.all_variants` takes care of jit, too. The problem was that the default value of alpha never changed. [74d17f0](https://github.com/google-deepmind/optax/pull/599/commits/74d17f0a76113d7dedd46f8e750064b4c0651c97) fixes this issue.